source/_posts/2016-10-24-gsoc-2016-port-nmatrix-to-jruby.markdown

```diff
@@ -0,0 +1,536 @@
+
+`NMatrix#[]` calls the `#xslice` operator which calls  `#get_slice` operator that use the stride to determine whether we are accessing a single element or multiple elements. If  there are multiple elements, `#dense_storage_get` returns an NMatrix object with the elements along the dimension.
+
+NMatrix-MRI differs from NMatrix-JRuby implementation as it makes sure that memory is properly utilized as the memory needs to be properly **garbage collected**.
+
+```ruby
+def xslice(args)
+  result = nil
+
+  if self.dim < args.length
+    raise(ArgumentError,"wrong number of arguments\
+       (#{args} for #{effective_dim(self)})")
+  else
+    result = Array.new()
+    slice = get_slice(@dim, args, @shape)
+    stride = get_stride(self)
+    if slice[:single]
+      if (@dtype == :object)
+        result = @s[dense_storage_get(slice,stride)]
+      else
+        s = @s.toArray().to_a
+        result = @s.getEntry(dense_storage_get(slice,stride))
+      end
+    else
+      result = dense_storage_get(slice,stride)
+    end
+  end
+  return result
+end
+```
+
+`NMatrix#[]=` calls the `#dense_storage_set` operator which calls `#get_slice` operator that use the stride to find out whether we are accessing a single element or multiple elements. If there are multiple
+ elements `#set_slice` recursively sets the elements of the matrix then returns an NMatrix object with the elements along the dimension.
+
+```ruby
+def dense_storage_set(slice, right)
+    stride = get_stride(self)
+    v_size = 1
+
+    if right.is_a?(NMatrix)
+      right = right.s.toArray.to_a
+    end
+
+    if(right.is_a?(Array))
+      v_size = right.length
+      v = right
+      if (dtype == :object)
+        # nm_register_values(reinterpret_cast<VALUE*>(v), v_size)
+      end
+
+      (0...v_size).each do |m|
+        v[m] = right[m]
+      end
+    else
+      v = [right]
+      if (@dtype == :object)
+        # nm_register_values(reinterpret_cast<VALUE*>(v), v_size)
+      end
+    end
+    if(slice[:single])
+      # reinterpret_cast<D*>(s->elements)[nm_dense_storage_pos(s, slice->coords)] = v;
+      pos = dense_storage_pos(slice[:coords],stride)
+      if @dtype == :object
+        @s[pos] = v[0]
+      else
+        @s.setEntry(pos, v[0])
+      end
+    else
+      v_offset = 0
+      dest = {}
+      dest[:stride] = get_stride(self)
+      dest[:shape] = shape
+      # dest[:elements] = @s.toArray().to_a
+      dense_pos = dense_storage_pos(slice[:coords],stride)
+      slice_set(dest, slice[:lengths], dense_pos, 0, v, v_size, v_offset)
+    end
+  end
+```
+
+## **Enumerators**
+
+NMatrix-MRI uses the C code for enumerating the elements of a matrix. However, the NMatrix-JRuby uses pure Ruby code. Currently, all the enumerators for dense matrices with real data-type have been implemented and they are properly functional. Enumerators for objects have not been currently implemented.
+
+```ruby
+def each_with_indices
```

Comment:
@mohawkjohn : I prefer providing code blocks. Is it mandatory to provide link to the code?